### PR TITLE
Implementation of Verifiers.cpp for linalg.matmul.

### PR DESCRIPTION
**Pull request by KoolJBlack**
Implements verifiers for the `LLVMGPUMatmulSimt` and `LLVMGPUMatmulTensorCore` pipelines with tests. 

Adds `Verifiers.cpp` to the `LLVMGPU` module for more thorough  verification.

### CODE REVIEW

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  if (secondLevelTileSizes[0] % tensorCoreSize[0] != 0 ||
+      secondLevelTileSizes[1] % tensorCoreSize[1] != 0 ||
+      secondLevelTileSizes[2] % tensorCoreSize[2] != 0) {
+    return op->emitOpError(
+               "tensorcore size doesn't factor into second level tile size "
+               "for ")
+           << pipelineName;
+  }
+
+  // For linalg::MatmulOp, verify the first level tile size divides the matmul
```
Comment (KoolJBlack) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:133]: @ThomasRaoux In our notes, we didn't list a shared memory size verifier for the TensorCore pipeline. Should there be one here as well?

Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:133]: Yes it should be the same. This is a hw/driver limit

**Comment by KoolJBlack** [comment]
in the `gpu_set_num_workgrous.mlir test`, there is a test case where the total workgroup size is 8: https://github.com/google/iree/blob/main/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir/#L102

This fails the total workgroup size check I've added to the verifier. @ThomasRaoux is this workgroup size correct?

**Comment by ThomasRaoux** [comment]
> in the `gpu_set_num_workgrous.mlir test`, there is a test case where the total workgroup size is 8: https://github.com/google/iree/blob/main/iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir/#L102
> 
> This fails the total workgroup size check I've added to the verifier. @ThomasRaoux is this workgroup size correct?

You can change the test, the test was meant to check that the annotation propagation works. I think at some point we might relax this rule but I'm not convinced it would work right now.

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
+#include "iree-dialects/Dialect/LinalgExt/Transforms/Passes.h"
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:1]: Please add the standard iree file header

**Comment by ThomasRaoux** [comment]
Looks like there are some test failures? Are you looking into it?

File: iree/compiler/Codegen/LLVMGPU/test/illegal_configuration.mlir [hunk 3]
```diff
@@ -32,38 +32,6 @@ hal.executable private @matmul_tensors {
 
 // -----
 
-#config = #iree_codegen.lowering.config<tile_sizes = [], native_vector_size = []>
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMGPU/test/illegal_configuration.mlir]: I am not sure we want to just remove the failing test?

**Comment by KoolJBlack** [comment]
I've fixed the illegal_configuration test. However, there are now two additional tests failing that I don't quite understand how it relates to this cl: 
`gpu_set_num_workgroups.mlir` and `nvvm_pipeline_test.mlir` both fail with:
```
CHECK-DAG: expected string not found in input
```

I can't actually reproduce these failures myself: 
```
The following tests passed:
        iree/compiler/Codegen/LLVMGPU/test/convert_to_nvvm.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/convert_to_rocdl.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/distribute_to_thread.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/distribute_wg_copy.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/gpu_set_num_workgroups.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/illegal_configuration.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/legalize.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/nvvm_pipeline_test.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/rocdl_pipeline_test.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/tensorcore_vectorization.mlir.test
        iree/compiler/Codegen/LLVMGPU/test/vectorization.mlir.test

100% tests passed, 0 tests failed out of 11

Label Time Summary:
iree/compiler/Codegen/LLVMGPU/test    =   6.66 sec*proc (11 tests)

Total Test time (real) =   6.72 sec
```
Not sure how it even relates to the verifier change. Any ideas?

**Review by ThomasRaoux — APPROVED** [review]

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    Operation *op, IREE::Codegen::LoweringConfigAttr loweringConfig,
+    IREE::Codegen::TranslationInfoAttr translationInfo,
+    ArrayRef<int64_t> workgroupSize) {
+  auto pipeline =
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulSimt;
+  StringRef pipelineName = stringifyEnum(pipeline);
+  if (workgroupSize.empty()) {
+    return op->emitOpError("expected workgroup size for GPU pipelines");
+  }
+  linalg::MatmulOp matmulOp = dyn_cast<linalg::MatmulOp>(op);
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:28]: any reason not to verify batch_matmul the same way it is done for the other pipeline? It is fine to do it in a separate PR but if it is the case please work on it next.

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+    if (firstLevelTileSizes[0] != 1) {
+      op->emitError("Received first tile dimension of ")
+          << firstLevelTileSizes[0] << " instead of 1 for " << pipelineName;
+    }
+    firstLevelTileSizes = {firstLevelTileSizes[1], firstLevelTileSizes[2],
+                           firstLevelTileSizes[3]};
+  } else {
+    return success(); // Only verify batched and unbatched matmul.
+  }
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:110]: I think this logic could be simplified. It could be something like:
```
if(!isa<linalg::MatmulOp, linalg::BatchMatmulOp>(op)) {
  return success();
}
inputType = matmulOp.getInputOperand(0)->get().getType();
lhsShape = getUntiledShape(matmulOp.getInputOperand(0)->get());
rhsShape = getUntiledShape(matmulOp.getInputOperand(1)->get());
firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
if(linalg::BatchMatmulOp batchMatmulOp =
                 dyn_cast<linalg::BatchMatmulOp>(op)) {
// drop batch dim, and verify batch tile size
...
}
```

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    lhsShape =
+        getUntiledShape(batchMatmulOp.getInputOperand(0)->get()).drop_front(1);
+    rhsShape =
+        getUntiledShape(batchMatmulOp.getInputOperand(1)->get()).drop_front(1);
+
+    // First tile dimensions should be 1 for batched, use remaining dimensions
+    // for comparisons.
+    firstLevelTileSizes = loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+    if (firstLevelTileSizes[0] != 1) {
+      op->emitError("Received first tile dimension of ")
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:103]: Can you add a test for this case?

**Review by ThomasRaoux — CHANGES_REQUESTED** [review]

File: iree/compiler/Codegen/LLVMGPU/Verifiers.cpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  // combined space.
+  unsigned bytesSize;
+  if (MemRefType type = inputType.dyn_cast<mlir::MemRefType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else if (RankedTensorType type = inputType.dyn_cast<RankedTensorType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else if (UnrankedTensorType type =
+                 inputType.dyn_cast<UnrankedTensorType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else {
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/LLVMGPU/Verifiers.cpp:73]: nit: unranked tensor type shouldn't be possible in this case. I would remove it.
